src/libraries/System.Text.Json/src/System.Text.Json.csproj

```diff
@@ -223,6 +224,12 @@
     <Compile Include="System\Text\Json\Writer\Utf8JsonWriter.WriteValues.UnsignedNumber.cs" />
     <Compile Include="System\TypeExtensions.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'">
+    <Compile Include="System\Text\Json\Reader\JsonReaderHelper.Intrinsics.cs" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' != '$(NetCoreAppCurrent)'">
```

Comment:
Note that many of our libraries that are inbox and OOB do this:

```xml
    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;netcoreapp3.0;net461</TargetFrameworks>
    <ExcludeCurrentNetCoreAppFromPackage>true</ExcludeCurrentNetCoreAppFromPackage>
```

This means that `NetCoreAppCurrent` is the build that goes in the runtimepack/shared framework. But in the NuGet package, only `netstandard2.0;netcoreapp3.0;net461` builds get packaged in there. So that means if someone references the OOB System.Text.Json v7 NuGet package, even on `net6.0`, they will use the `netcoreapp3.0` version from the package, because that's the highest version available.

So I really think these conditions should be `'$(TargetFramework)' == '$(NetCoreAppCurrent)' or '$(TargetFramework)' == 'netcoreapp3.0'` - Include `.Intrinsics`.  Else - include `.NoIntrinsics`.

Reply:
The intrinsics will break on netcoreapp3.0 though because they include the arm variants? Should I include `net5.0` and treat that as the netcoreapp3.0 in your example?

Reply:
This probably needs the experts.

@ericstj @steveharter @layomia - how do you want to handle Arm intrinsics in STJ? Do you want to ship a `net5.0` TFM in the OOB package?

Reply:
Could type forward, but not sure how that works either 😅

Reply:
Type forwards wouldn't work here. I assume to support Arm we will need a new TFM in the package. But I think that's up to the System.Text.Json owners to decide. I'm just a fly-by reviewer 😄

Reply:
The point of the OOB packages like this is to make things work, not to make them super fast. I do not think we need to bother with adding a new build to the OOB package.

Reply:
Was this comment that concerned me:

> So that means if someone references the OOB System.Text.Json v7 NuGet package, even on `net6.0`, they will use the `netcoreapp3.0` version from the package, because that's the highest version available.

Reply:
that's true but I guess the concern here would be the possible perf regression in the future, for example with this scenario:
- You have an app that targets net6.0 (uses inbox version of S.T.J which has the optimization)
- We ship a new package in 6.x o 7.0 timeframe that doesn't have netcoreappcurrent in it (so it doesn't have the fast version)
- Your app adds a dependency on package Foo, which has a indirect dependency to S.T.J version 6.x/7.0.
- Now your app will use the S.T.J from the package due to conflict resolution, which presumably will be the nca3.0 one and your app will have a perf regression.

In general I agree with @jkotas point, but the scenario above is something to keep in mind

Reply:
The OOB should never downgrade the functionality in the framework.  It will take precedent over the framework assembly, that's what lets us to continue to evolve the API.  We don't want the package version to substantially degrade the inbox version.  In general my rule is if we need `ifdefs` or conditional source compilation then it needs to be in the package.

I know @tannergooding had talked about having downlevel representation of the intrinsics to allow their use from older frameworks, not sure if that's a thing or not yet.

Reply:
The issue is when the OOB package gets pulled in to your app (maybe because one of your references updated to the latest STJ), now your app suddenly gets slower. No other changes happened except updating dependency versions, and your app regresses.

Reply:
Was this comment that concerned me:

> So that means if someone references the OOB System.Text.Json v7 NuGet package, even on `net6.0`, they will use the `netcoreapp3.0` version from the package, because that's the highest version available.

Am ok with the faster stuff only applying to `net6.0` since its also LTS, so there is less reason for backfilling; but if referencing it from NuGet would override? 🤷‍♂️

e.g. how does it resolve if a .NETstd lib refers to the NuGet?

Reply:
> The issue is when the OOB package gets pulled in to your app (maybe because one of your references updated to the latest STJ), now your app suddenly gets slower. 

The OOB packages make your app bigger and slower by design. For example, the OOB package does not come with R2R compiled code, so your app will startup slower when you reference the OOB package.

The bottom line is: If you want to depend on latest version, you better update the runtime too.

Reply:
So not worth worrying about; until there is a mid-release release, as it were?

Reply:
> The OOB packages make your app bigger and slower by design. For example, the OOB package does not come with R2R compiled code, so your app will startup slower when you reference the OOB package.

Not necessarily, I might build self-contained and run cross-gen myself.  Or I might be running on an aot runtime.

>  If you want to depend on latest version, you better update the runtime too.

That's not always an option.  Consider what happens in .NET 7.0.  We're going to add API to this library and I bet the Azure SDK (or some other libs) will want to use it.  At that point any 6.0 app that uses Azure SDK will have no choice but to use the package version.

I'd prefer we keep a simple rule here.  If the configuration differs in conditions/ifdefs from the packaged configurations then it should also be included in the package.

Reply:
> At that point any 6.0 app that uses Azure SDK will have no choice but to use the package version.

Yes, it will continue to work - with small perf hit - even if we keep the status quo.

It is ok with me if we want build extra targets to ensure that newer OOB packages on downlevel platforms always get all performance tweaks possible.

There are other assemblies with this problem. For example,  
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj uses ARM intrisics inbox, but there is no OOB build with ARM intrinsics. I assume that we would want to fix all these.

Reply:
>  Consider what happens in .NET 7.0. We're going to add API to this library and I bet the Azure SDK (or some other libs) will want to use it. At that point any 6.0 app that uses Azure SDK will have no choice but to use the package version.

Sounds like an issue to resolve before next year's 7.0 release?  😉

Reply:
Nothing gets resolved by that point.  New API in the package means people will use that new API, so consuming the package that's newer than the framework is super common.  Folks shouldn't have to pay a penalty when they do that.  My position on this is don't use `ExcludeCurrentNetCoreAppFromPackage` if there is anything different about the `NetCoreAppCurrent` build.  Don't split hairs on something saying it's OK because it's only a small perf fix.  You might be able to convince yourself of that, but what about the next person that doesn't notice it's excluded?  What about the customer that cares about that perf-fix.  
The package growth problem is less of an issue now that we're starting to drop out of support frameworks from packages thanks to precedent that @ViktorHofer is setting.
Let's keep our rules simple: if you if'def/conditionally compile, then don't exclude from package.

Reply:
Regarding the initial feedback, please don't set `ExcludeCurrentNetCoreAppFromPackage` anymore as we don't want to exclude the latest .NETCoreApp asset in packages anymore. See https://github.com/dotnet/runtime/pull/53439 for more details.

As @ericstj mentioned, in the past we excluded specific tfms - even though they were shipping in some form - from our packages but based on the new package support policy we don't carry along unsupported .NETCoreApp assets along anymore which bounds the number of .NETCoreApp assets in a package.